modules/ctl/src/main/scala/free/interp.scala

```diff
@@ -0,0 +1,119 @@
+package gem.ctl
+package free
+
+import scalaz._, Scalaz._
+import scalaz.effect._
+
+import gem.ctl.low.io._
+import gem.ctl.free.ctl._
+
+/** An interpreter for the control language. */
+object interpreter {
+
+  /** Configuration for the interpreter. */
+  trait Config {
+    def verbose: Boolean
+    def server:  Server
+  }
+
+  case class InterpreterState(indentation: Int, machineHostCache: Map[Host.Machine, String]) {
+    def indent  = copy(indentation = indentation + 1)
+    def outdent = copy(indentation = indentation - 1)
+  }
+  object InterpreterState {
+    val initial = InterpreterState(0, Map.empty)
+  }
+
+  /**
+   * Construct an interpreter of `CtlIO` given a `Config` and an `IORef` for the initial state. We
+   * carry our state in an `IORef` because it needs to be visible from multiple threads.
+   */
+  def interpreter(c: Config, state: IORef[InterpreterState]) = λ[CtlOp ~> EitherT[IO, Int, ?]] {
```

Comment:
From by biased background from Haskell, this looks like a normal type signature that you'd like to easily express in any pure FP language. So I'm glad that kind projectors makes this cleaner, because without them something that should be straightforward is made more difficult because of syntactic mess.

But this this discussion, as many others before, seems to be leading to do we use Scala as a more modern Java with some styling borrowed from FP or are we invested on truly using the Scala as a true FP language.